tests/raftstore/test_split_region.rs

```diff
@@ -302,4 +302,79 @@ fn test_node_split_overlap_snapshot() {
 fn test_server_split_overlap_snapshot() {
     let mut cluster = new_server_cluster(0, 3);
     test_split_overlap_snapshot(&mut cluster);
-}
\ No newline at end of file
+}
+
+fn test_apply_new_version_snapshot<T: Simulator>(cluster: &mut Cluster<T>) {
+    // truncate the log quickly so that we can force sending snapshot.
+    cluster.cfg.store_cfg.raft_log_gc_tick_interval = 20;
+    cluster.cfg.store_cfg.raft_log_gc_limit = 2;
+
+    // We use three nodes 1, 2, 3 for this test.
+    cluster.bootstrap_region().expect("");
+    cluster.start();
+
+    // guarantee node 1 is leader
+    cluster.transfer_leader(1, util::new_peer(1, 1));
+    cluster.must_put(b"k0", b"v0");
+    assert_eq!(cluster.leader_of_region(1), Some(util::new_peer(1, 1)));
+
+    let pd_client = cluster.pd_client.clone();
+
+    // isolate node 3 for region 1.
+    cluster.hook_transport(IsolateRegionStore::new(1, 3));
+    cluster.must_put(b"k1", b"v1");
+
+    let region = pd_client.get_region(b"").unwrap();
+
+    // split (-inf, +inf) -> (-inf, k2), [k2, +inf]
+    cluster.must_split(&region, b"k2");
+    cluster.must_put(b"k2", b"v2");
+
+    // node 1 and node 2 must have k2, but node 3 must not.
+    for i in 1..3 {
+        let engine = cluster.get_engine(i);
+        util::must_get_equal(&engine, b"k2", b"v2");
+    }
+
+    let engine = cluster.get_engine(3);
+    util::must_get_none(&engine, b"k2");
+
+    for _ in 0..100 {
+        // write many logs to force log GC for region 1 and region 2.
+        cluster.get(b"k1").unwrap();
+        cluster.get(b"k2").unwrap();
+    }
+
+    let region2 = pd_client.get_region(b"k2").unwrap();
+    // split [k2, +inf) -> [k2, k3), [k3, +inf)
+    cluster.must_split(&region2, b"k3");
```

Comment:
For more test, and this covers the bug for applying snapshot with uninitialized peer.